docs/source/notes/cuda.rst

```diff
@@ -635,6 +635,207 @@ of the alloc/free functions that match the signatures specified above.
+be called internally on deletion of the pool, hence returning all the memory to the system.
+
+.. code:: python
+
+   del tensor, del pool
+
+
+The following :meth:`torch.cuda.MemPool.use_count` and :meth:`torch.cuda.MemPool.snapshot`
+APIs can be used for debugging purposes:
+
+.. code:: python
+
+   pool = torch.cuda.MemPool(allocator)
+
+   # pool's use count should be 1 at this point as MemPool object
+   # holds a reference
+   assert pool.use_count() == 1
+
+   nelem_1mb = 1024 * 1024 // 4
+
+   with torch.cuda.use_mem_pool(pool):
+       out_0 = torch.randn(nelem_1mb, device="cuda")
+
+       # pool's use count should be 2 at this point as use_mem_pool
+       # holds a reference
+       assert pool.use_count() == 2
+
+   # pool's use count should be back to 1 at this point as use_mem_pool
+   # released its reference
+   assert pool.use_count() == 1
+
+   with torch.cuda.use_mem_pool(pool):
+       # pool should have 1 segment since we made a small allocation (1 MB)
+       # above and so the CUDACachingAllocator packed it into a 2 MB buffer
+       assert len(pool.snapshot()) == 1
+
+       out_1 = torch.randn(nelem_1mb, device="cuda")
+
+       # pool should still have 1 segment since we made another small allocation
+       # (1 MB) that got packed into the existing 2 MB buffer
+       assert len(pool.snapshot()) == 1
+
+       out_2 = torch.randn(nelem_1mb, device="cuda")
+
+       # pool now should have 2 segments since the CUDACachingAllocator had
+       # to make a new 2 MB buffer to accomodate out_2
+       assert len(pool.snapshot()) == 2
+
+
+.. note::
+
+   * :class:`torch.cuda.MemPool` holds a reference to the pool. When you use the
+     :class:`torch.cuda.use_mem_pool` context manager, it will also acquire another reference
+     to the pool. On exit of the context manager, it will release its reference. After that,
+     ideally it should only be tensors holding references to the pool. Once the tensors release
+     their references, the use count of the pool will be 1, reflecting that only the
+     :class:`torch.cuda.MemPool` object is holding a reference. Only at that point, can the memory
+     held by the pool be returned to the system when the pool's destructor is called using
+     ``del``.
+   * :class:`torch.cuda.MemPool` doesn't currently support ``expandable_segments`` mode of
+     CUDACachingAllocator.
+   * `NCCL has specific requirements`_ for a buffer to be compatible with NVLS reductions.
+     These requirements can be broken in a dynamic workload, for instance, the buffer being
+     sent to NCCL by the CUDACachingAllocator might be split and hence, not correctly aligned.
+     In those cases, NCCL can use a fallback algorithm instead of NVLS.
+   * Allocators like ``ncclMemAlloc`` can use more memory than requested, due to alignment
+     requirements (``CU_MULTICAST_GRANULARITY_RECOMMENDED``, ``CU_MULTICAST_GRANULARITY_MINIMUM``),
+     and can cause your workload to run out of memory.
```

Comment:
Thank you for documenting this!

If I give `ncclMemAlloc` backed allocator to MemPool, and I ask for a tensor of say 4 MB, would MemPool:
- call `ncclMemAlloc` with 4 MB ask exactly, or 
- call `ncclMemAlloc` with a large-segment ask (like 512 MB) and slice a 4 MB region as return?

Reply:
Sorry, why would it use 512MB segment ask? In the caching allocator, small segments are rounded up to 2MB, but large segments are more or less called as is, only rounded up to 2MB bytes or so, so I don't see conditions under which 4MB request would result in 512 MB allocation. ncclMemAlloc itself is allocating at `CU_MEM_ALLOC_GRANULARITY_RECOMMENDED` which for H100 is 2MB, so also nowhere near 512 MB